### ISSUE TITLE
outgoing rich text signed pgp/mime message broken

### ISSUE THREAD

**Issue by tomholub** [FlowCrypt/flowcrypt-browser#3157]
possibly caused by #3127 - at least it worked on last version

it only affects rich text, which is only allowed on firefox

To reproduce:
1) use firefox
2) using send button chevron inco switch to "Rich text" and disable "encrypt" so you have a signed + rich text message
3) add text, you can make some part bold with ctrl+b just to make sure, then attach an image
4) view the sent message using FlowCrypt
5) view the sent message using Thunderbird (I haven't tried this)

![image](https://user-images.githubusercontent.com/6306961/99191460-c1688f00-2764-11eb-9b8f-5d9bcb415abb.png)

![image](https://user-images.githubusercontent.com/6306961/99191427-92521d80-2764-11eb-959b-325d521621df.png)

**Comment by tomholub** [comment]
Here is the message it produced

[refactor - signed rich.txt](https://github.com/FlowCrypt/flowcrypt-browser/files/5542790/refactor.-.signed.rich.txt)

**Comment by rrrooommmaaa** [comment]
broken.txt
-- this is how it looks in gmail.
![image](https://user-images.githubusercontent.com/23335944/99654492-7489fe80-2a6b-11eb-8a05-9fd4e4953dc1.png)

in FlowCrypt "inbox":
![image](https://user-images.githubusercontent.com/23335944/99654566-8f5c7300-2a6b-11eb-9fea-47ab00916b8a.png)

So the error is that image is recognized as "plain file" in "inbox"?

**Comment by tomholub** [comment]
The error is that on my end, I'm actually seeing this in Gmail:

![image](https://user-images.githubusercontent.com/6306961/99657063-924c6900-2a55-11eb-88b6-142fe29da4fb.png)

**Comment by tomholub** [comment]
It's interesting that it rendered correctly for you, but not for me. Tried on both Chrome and Firefox.

![image](https://user-images.githubusercontent.com/6306961/99657348-f5d69680-2a55-11eb-8d3f-0ba26176b277.png)

**Comment by tomholub** [comment]
I just noticed - you were looking at the message in the sent:folder.

I'm looking at the received message, which is the one I exported.

Try to import the received message, then you should be able to see the problem.

**Comment by rrrooommmaaa** [comment]
@tomholub I see that the `broken.txt` file you attached has different delimeters as compared with Flowcrypt's "sent". So it is sort of "before processing by google", correct?

**Comment by tomholub** [comment]
Hm. This used to work. Can you please test sending out a message from a little older firefox addon? For example

https://s3-us-west-2.amazonaws.com/cryptup-release/firefox/cryptup-io-firefox-7.9.5.xpi or even older.

I want to make sure that it's not us that started sending messages a bit differently, causing this Gmail mechanism to get triggered. I remember writing the sending code in a very specific manner to avoid Gmail API malforming the message for pgp/mime signed messages, and I wonder if something in that code changed on our end.

**Comment by rrrooommmaaa** [comment]
@tomholub Of course, modified message can't be verified if it is edited (boundaries have changed). So it should be not parse error, but verification error. I'm setting up tests for original and modified message.

**Comment by tomholub** [comment]
I'm not so worried if it's parse or verification error.

I wonder if anything changed about how we encode the message, which in turn caused google to start changing the boundaries that it didn't before.

**Comment by rrrooommmaaa** [comment]
7.9.5 produced the same error.
I compared "rich text from prev version signed" and original "refactor - signed rich" from "Sent" folder -- they are identical in structure.
![image](https://user-images.githubusercontent.com/23335944/99778064-3ac8fe80-2b24-11eb-90fc-40cbbbba95c9.png)

The deformed message on arrival doesn't have `multipart/signed`, it is re-formatted as `multipart/mixed` (by Gmail API?) as per #3032

**Comment by tomholub** [comment]
Excellent, thank you. I'll report it to them.

**Comment by tomholub** [comment]
I reported this also at https://partnerissuetracker.corp.google.com/issues/171715824

**Comment by tomholub** [comment]
Thank you - unassigned - waiting for Google.

**Comment by tomholub** [comment]
For context this is what I reported

--------------------------

Me:

short description: Structure of PGP/MIME messages should be preserved when sending through Gmail API (users.messages.send)

This bug affects several large organizations who use encrypted email, including Google's strategic enterprise customer.

Steps to reproduce:
1) compose a pgp/mime encrypted message (pgp/mime signed works ok)
2) send it through Gmail API to another account (Gmail or otherwise)
3) open the message on the receiving end (not on sender's side)
4) message will be malformed as described below

The sent message had the following structure:
```
Content-Type: multipart/encrypted; protocol="application/pgp-encrypted"; boundary="----sinikael-?=_1-16031047614490.8483156050873946"

------sinikael-?=_1-16031047614490.8483156050873946
Content-Type: application/pgp-encrypted; name=
Content-Description: PGP/MIME version identification
Content-Disposition: attachment
X-Attachment-Id: f_uyeXnnnxBHMzPjkQpzUnKlwciCoyZB@flowcrypt
Content-Id: <f_uyeXnnnxBHMzPjkQpzUnKlwciCoyZB@flowcrypt>
Content-Transfer-Encoding: base64


------sinikael-?=_1-16031047614490.8483156050873946
Content-Type: application/octet-stream; name=encrypted.asc
Content-Description: OpenPGP encrypted message
Content-Disposition: inline; filename=encrypted.asc
X-Attachment-Id: f_AcohxKCOvwZLUPQEMgsUwfzMzfpfWS@flowcrypt
Content-Id: <f_AcohxKCOvwZLUPQEMgsUwfzMzfpfWS@flowcrypt>
Content-Transfer-Encoding: base64

<snip>
------sinikael-?=_1-16031047614490.8483156050873946
Content-Type: application/pgp-keys; name=0x8C826EA5BCD44F29.asc
Content-Disposition: attachment; filename=0x8C826EA5BCD44F29.asc
X-Attachment-Id: f_btAAVzsmZSFAgNRdjjBNYQJegGCNjB@flowcrypt
Content-Id: <f_btAAVzsmZSFAgNRdjjBNYQJegGCNjB@flowcrypt>
Content-Transfer-Encoding: base64

<snip>
------sinikael-?=_1-16031047614490.8483156050873946--
```
The following message was delivered and pulled on the receiving end:
```
Content-Type: multipart/mixed; boundary="000000000000e5587e05b203e91d"

--000000000000e5587e05b203e91d
Content-Type: text/plain; charset="UTF-8"

<snip>

--000000000000e5587e05b203e91d
Content-Type: application/pgp-encrypted; name=""
Content-Disposition: attachment
Content-Transfer-Encoding: base64
Content-ID: <f_uyeXnnnxBHMzPjkQpzUnKlwciCoyZB@flowcrypt>
X-Attachment-Id: f_uyeXnnnxBHMzPjkQpzUnKlwciCoyZB@flowcrypt


--000000000000e5587e05b203e91d
Content-Type: application/pgp-keys; name="0x8C826EA5BCD44F29.asc"
Content-Disposition: attachment; filename="0x8C826EA5BCD44F29.asc"
Content-Transfer-Encoding: base64
Content-ID: <f_btAAVzsmZSFAgNRdjjBNYQJegGCNjB@flowcrypt>
X-Attachment-Id: f_btAAVzsmZSFAgNRdjjBNYQJegGCNjB@flowcrypt

<snip>
--000000000000e5587e05b203e91d--
```

In summary, the following changes caused the incompatibility:
1) `Content-Type: multipart/encrypted; protocol="application/pgp-encrypted";` changed to `Content-Type: multipart/mixed;`
2) attachment with `Content-Type: application/octet-stream; name=encrypted.asc` lost all headers and changed to `Content-Type: text/plain; charset="UTF-8"`, also changed order
3) attachment with `Content-Type: application/pgp-encrypted; name=` lost headers, and changed order

Full messages attached.

The appropriate RFC is at https://tools.ietf.org/html/rfc3156#section-4

I'm available to provide more details if needed.

Cheers,
Tom

----------------------

Google:

> Thank you for filing this issue. I forwarded it internally and it is being investigated now.

-----------------------

Me:

Thank you. In the last few days, we have noticed similar issue newly started happening also on pgp/mime signed messages that didn't happen before.

It's not clear at what point did this newly break - we first observed it on Nov 15.
